cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -1416,88 +1414,21 @@ private AppFabricServiceStatus start(final Id.Program id, ProgramType type,
-                                                             userArguments, debug));
-
-      final ProgramController controller = runtimeInfo.getController();
-      final String runId = controller.getRunId().getId();
-
-      if (type != ProgramType.MAPREDUCE) {
-        // MapReduce state recording is done by the MapReduceProgramRunner
-        // TODO [JIRA: CDAP-2013] Same needs to be done for other programs as well
-        controller.addListener(new AbstractListener() {
-          @Override
-          public void init(ProgramController.State state, @Nullable Throwable cause) {
-            // Get start time from RunId
-            long startTimeInSeconds = RunIds.getTime(controller.getRunId(), TimeUnit.SECONDS);
-            if (startTimeInSeconds == -1) {
-              // If RunId is not time-based, use current time as start time
-              startTimeInSeconds = TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis());
-            }
-            store.setStart(id, runId, startTimeInSeconds);
-            if (state == ProgramController.State.COMPLETED) {
-              completed();
-            }
-            if (state == ProgramController.State.ERROR) {
-              error(controller.getFailureCause());
-            }
-          }
-
-          @Override
-          public void completed() {
-            store.setStop(id, runId, TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis()),
-                          ProgramController.State.COMPLETED.getRunStatus());
-          }
-
-          @Override
-          public void killed() {
-            store.setStop(id, runId, TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis()),
-                          ProgramController.State.KILLED.getRunStatus());
-          }
-
-          @Override
-          public void suspended() {
-            store.setSuspend(id, runId);
-          }
-
-          @Override
-          public void resuming() {
-            store.setResume(id, runId);
-          }
-
-          @Override
-          public void error(Throwable cause) {
-            LOG.info("Program stopped with error {}, {}", id, runId, cause);
-            store.setStop(id, runId, TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis()),
-                          ProgramController.State.ERROR.getRunStatus());
-          }
-        }, Threads.SAME_THREAD_EXECUTOR);
-      }
-      return AppFabricServiceStatus.OK;
-    } catch (DatasetInstantiationException e) {
-      return new AppFabricServiceStatus(HttpResponseStatus.UNPROCESSABLE_ENTITY, e.getMessage());
+      ProgramRuntimeService.RuntimeInfo runtimeInfo = lifecycleService.start(id, type, sysArgs, userArgs, debug);
+      return (runtimeInfo != null) ? AppFabricServiceStatus.OK : AppFabricServiceStatus.INTERNAL_ERROR;
     } catch (Throwable throwable) {
       LOG.error(throwable.getMessage(), throwable);
-      if (throwable instanceof FileNotFoundException) {
+      if (throwable instanceof ProgramNotFoundException) {
```

Comment:
should just have a separate catch for this exception instead of checking the type here. Also be sure not to log an error for this case.